src/cache/intern.rs

```diff
@@ -0,0 +1,78 @@
+use cfg_if::cfg_if;
+
+
+cfg_if! {
+    if #[cfg(feature = "enable-interning")] {
+        use std::thread_local;
+        use std::string::String;
+        use std::borrow::ToOwned;
+        use std::cell::RefCell;
+        use std::collections::HashMap;
+        use crate::JsValue;
+        use crate::convert::IntoWasmAbi;
+
+        struct Cache {
+            entries: RefCell<HashMap<String, JsValue>>,
+        }
+
+        thread_local! {
+            static CACHE: Cache = Cache {
+                entries: RefCell::new(HashMap::new()),
+            };
+        }
+
+        /// This returns the raw index of the cached JsValue, so you must take care
+        /// so that you don't use it after it is freed.
+        pub(crate) fn unsafe_get_str(s: &str) -> Option<<JsValue as IntoWasmAbi>::Abi> {
+            CACHE.with(|cache| {
+                let cache = cache.entries.borrow();
+
+                cache.get(s).map(|x| x.into_abi())
+            })
+        }
+
+        fn intern_str(key: &str) {
+            CACHE.with(|cache| {
+                let mut cache = cache.entries.borrow_mut();
+
+                // Can't use `entry` because `entry` requires a `String`
+                if !cache.contains_key(key) {
+                    cache.insert(key.to_owned(), JsValue::from(key));
```

Comment:
I think that we're going to want some form of eviction for this, because otherwise the interning feature may not be too useful as the Rust memory would just grow without bound. Could we perhaps have something like a 1MB string limit or maybe a maximum amount limit for interning?

(it may also be good to skip interning huge strings over a threshold as well)

Reply:
Do you think that will be an issue in practice? Strings are only interned if the user explicitly and manually calls the `intern` function. They are never automatically interned.

And the documentation for `intern` makes it clear that you should only use `intern` if there's a high probability that the string will be sent to JS multiple times.

So, what if the user actually *wants* to intern a large string which is sent to JS multiple times?

Reply:
Oh wow I've been thinking about this wrong the entire time, I thought everything automatically went through interning!

Nah in that case this is fine for now.

Reply:
Yeah, it *used* to automatically cache, but when that caused performance issues [I changed it to manual caching](https://github.com/rustwasm/wasm-bindgen/pull/1612#issuecomment-504700943).